tests/test_main.py

```diff
@@ -95,32 +99,43 @@ def on_save(fh, res, conf):
     def test_input_invalid_json(self):
         test_input_json = """"foo": "bar", "baz": "bliff" }"""
 
-        def on_save(fh, res, conf):
+        def on_save(self, fh, res):
             pass
 
-        def on_done(res):
-            res = res.__dict__
-            self.assertEqual(res["status"], 400)
+        def on_done(self):
+            res = self.res.__dict__
+            assert res["status"] == 400
             sys.exit(1)
 
-        p = proxy.Proxy(self.conf, proxy.Req(), on_save, on_done)
+        p = proxy.Proxy(self.conf_path, proxy.Req())
+
+        # patching on_save and on_done for tests
+
+        p.on_done = types.MethodType(on_done, p)
+        p.on_save = types.MethodType(on_save, p)
```

Comment:
regarding the use of `types.MethodType` here, can't we do this method patching with `unittest.mock.patch.object`?

Reply:
we discussed this just now synchronously, we can do this patching with `unittest.mock.patch.object`, but since `types.MethodType` is more readable and since we don't need to worry about setting `Proxy.on_done`'s methods back to their original values after the test run (the `Proxy` object `p` is only used in this test), let's leave it be